File: src/services/file_watcher.rs
use notify::Watcher;
use sqlx::PgPool;
use std::path::Path;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::services::cache::CacheService;

pub struct FileWatcher {
    running: Arc<Mutex<bool>>,
    stop_tx: Arc<Mutex<Option<tokio::sync::oneshot::Sender<()>>>>,
}

impl FileWatcher {
    pub fn new() -> Self {
        Self {
            running: Arc::new(Mutex::new(false)),
            stop_tx: Arc::new(Mutex::new(None)),
        }
    }

    pub async fn start(&self, pool: &PgPool, cache: &CacheService) {
        // Check if enabled
        let enabled: Option<String> =
            sqlx::query_scalar("SELECT value FROM system_config WHERE key = 'file_watcher_enabled'")
                .fetch_optional(pool)
                .await
                .ok()
                .flatten();

        if enabled.as_deref() == Some("false") {
            tracing::info!("[FileWatcher] Disabled by config");
            return;
        }

        let libs = crate::models::library::get_all_libraries(pool).await.unwrap_or_default();
        if libs.is_empty() {
            tracing::info!("[FileWatcher] No libraries to watch");
            return;
        }

        // Collect all library paths with their library info
        let mut watch_paths: Vec<(String, String, String)> = Vec::new(); // (path, library_id, collection_type)
        for lib in &libs {
            for p in &lib.paths {
                if p.is_empty() || !Path::new(p).exists() {
                    continue;
                }
                if is_remote_mount(p) {
                    tracing::info!("[FileWatcher] Skipping remote mount: {p}");
                    continue;
                }
                watch_paths.push((p.clone(), lib.id.to_string(), lib.collection_type.clone()));
            }
        }

        if watch_paths.is_empty() {
            tracing::info!("[FileWatcher] No local paths to watch");
            return;
        }

        let (stop_tx, stop_rx) = tokio::sync::oneshot::channel::<()>();
        *self.stop_tx.lock().await = Some(stop_tx);
        *self.running.lock().await = true;

        let pool = pool.clone();
        let cache = cache.clone();

        tokio::spawn(async move {
            run_watcher(pool, cache, watch_paths, stop_rx).await;
        });
    }

    pub async fn stop(&self) {
        if let Some(tx) = self.stop_tx.lock().await.take() {
            let _ = tx.send(());
        }
        *self.running.lock().await = false;
        tracing::info!("[FileWatcher] Stopped");
    }

    pub async fn restart(&self, pool: &PgPool, cache: &CacheService) {
        self.stop().await;
        self.start(pool, cache).await;
    }
}

async fn run_watcher(
    pool: PgPool,
    cache: CacheService,
    watch_paths: Vec<(String, String, String)>,
    mut stop_rx: tokio::sync::oneshot::Receiver<()>,
) {
    let (tx, rx) = std::sync::mpsc::channel();

    let mut watcher = match notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| {
        if let Ok(event) = res {
            let _ = tx.send(event);
        }
    }) {
        Ok(w) => w,
        Err(e) => {
            tracing::error!("[FileWatcher] Failed to create watcher: {e}");
            return;
        }
    };

    // Watch all paths recursively
    let mut watched = 0;
    for (path, _, _) in &watch_paths {
        if let Err(e) = watcher.watch(Path::new(path), notify::RecursiveMode::Recursive) {
            tracing::warn!("[FileWatcher] Cannot watch {path}: {e}");
        } else {
            watched += 1;
        }
    }

    tracing::info!("[FileWatcher] Watching {watched} paths for changes");

    // Build a lookup: path prefix → (library_id, collection_type)
    let path_map: Vec<(String, String, String)> = watch_paths;

    // Process events in a loop
    loop {
        tokio::select! {
            _ = &mut stop_rx => {
                tracing::info!("[FileWatcher] Stop signal received");
                break;
            }
            _ = tokio::time::sleep(std::time::Duration::from_millis(500)) => {
                // Drain all pending events
                let mut events: Vec<notify::Event> = Vec::new();
                while let Ok(event) = rx.try_recv() {
                    events.push(event);
                }
                if events.is_empty() { continue; }

                // Deduplicate paths
                let mut created_paths = std::collections::HashSet::new();
                let mut removed_paths = std::collections::HashSet::new();

                for event in &events {
                    match event.kind {
                        notify::EventKind::Create(_) => {
                            for p in &event.paths {
                                created_paths.insert(p.to_string_lossy().to_string());
                            }
                        }
                        notify::EventKind::Remove(_) => {
                            for p in &event.paths {
                                removed_paths.insert(p.to_string_lossy().to_string());
                            }
                        }
                        notify::EventKind::Modify(notify::event::ModifyKind::Name(_)) => {
                            // Rename: treat as remove old + create new
                            for p in &event.paths {
                                let ps = p.to_string_lossy().to_string();
                                if p.exists() {
                                    created_paths.insert(ps);
                                } else {
                                    removed_paths.insert(ps);
                                }
                            }
                        }
                        _ => {}
                    }
                }

                // Handle removals: delete from DB
                for removed in &removed_paths {
                    handle_file_removed(&pool, removed).await;
                }

                // Handle creations: trigger scan for the relevant library
                let mut libs_to_scan = std::collections::HashSet::new();
                for created in &created_paths {
                    for (prefix, lib_id, _ctype) in &path_map {
                        if created.starts_with(prefix) {
                            libs_to_scan.insert(lib_id.clone());
                            break;
                        }
                    }
                }

                if !libs_to_scan.is_empty() {
                    for lib_id in &libs_to_scan {
                        // Find library info
                        if let Some((_, lid, ctype)) = path_map.iter().find(|(_, id, _)| id == lib_id) {
                            let lib = crate::models::library::get_all_libraries(&pool).await.unwrap_or_default()
                                .into_iter().find(|l| l.id.to_string() == *lid);
                            if let Some(lib) = lib {
                                tracing::info!("[FileWatcher] Changes detected, rescanning: {}", lib.name);
                                let tracker = crate::services::scan_progress::ScanProgressTracker::new();
                                crate::services::scanner::scan_library(
                                    &pool, &cache, &tracker,
                                    &lib.id.to_string(), &lib.collection_type, &lib.paths, &lib.name,
                                ).await;
                            }
                        }
                    }
                }
            }
        }
    }
}

async fn handle_file_removed(pool: &PgPool, file_path: &str) {
    // Check if this path is a known item
    let row: Option<(uuid::Uuid, String)> = sqlx::query_as(
        "SELECT id, type FROM items WHERE file_path = $1 LIMIT 1"
    ).bind(file_path).fetch_optional(pool).await.ok().flatten();

    if let Some((id, item_type)) = row {
        sqlx::query("DELETE FROM items WHERE id = $1").bind(id).execute(pool).await.ok();
        tracing::info!("[FileWatcher] Removed {item_type} from DB: {file_path}");

        // Cleanup empty parents
        if item_type == "Episode" {
            // Remove empty seasons
            sqlx::query(
                "DELETE FROM items WHERE type = 'Season' AND NOT EXISTS (SELECT 1 FROM items e WHERE e.parent_id = items.id)"
            ).execute(pool).await.ok();
            // Remove empty series
            sqlx::query(
                "DELETE FROM items WHERE type = 'Series' AND NOT EXISTS (SELECT 1 FROM items c WHERE c.parent_id = items.id)"
            ).execute(pool).await.ok();
        }
    }
}

fn is_remote_mount(dir_path: &str) -> bool {
    let output = std::process::Command::new("df")
        .args(["-T", dir_path])
        .output();

    if let Ok(out) = output {
        let text = String::from_utf8_lossy(&out.stdout);
        if let Some(line) = text.lines().nth(1) {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if let Some(fs_type) = parts.get(1) {
                let ft = fs_type.to_lowercase();
                return ft.starts_with("fuse")
                    || ["nfs", "nfs4", "cifs", "smb", "smbfs", "9p", "sshfs"]
                        .contains(&ft.as_str());
            }
        }
    }
    false
}

File: internal/services/file_watcher.go
package services

import (
	"context"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
	"github.com/jackc/pgx/v5/pgxpool"

	"fyms/internal/repository"
)

// FileWatcher 是 ingest 事件流的 fsnotify producer。
// Phase 1 之前它自己 debounce 后调 ScanLibrary 触发全库扫,大库一次改名要几分钟;
// 现在它只把 fsnotify 原始事件翻译成 IngestEvent 丢给 IngestWorker,
// 500ms 窗口的 Rename 合并、目录级 Delete、路径归属判断都在 worker 侧做。
type FileWatcher struct {
	mu        sync.Mutex
	running   bool
	stopCh    chan struct{}
	ingest    *IngestWorker
	refreshes *RefreshScheduler
}

func NewFileWatcher(ingest *IngestWorker, refreshes *RefreshScheduler) *FileWatcher {
	return &FileWatcher{ingest: ingest, refreshes: refreshes}
}

func (fw *FileWatcher) Start(ctx context.Context, pool *pgxpool.Pool, cache *CacheService) {
	enabled, ok, _ := repository.NewSystemConfigRepository(pool).GetString(ctx, "file_watcher_enabled")
	if ok && enabled == "false" {
		slog.Info("[FileWatcher] Disabled by config")
		return
	}

	libs, err := repository.NewLibraryRepository(pool).ListLibrariesForWatcher(ctx)
	if err != nil {
		slog.Error("[FileWatcher] Failed to get libraries", "error", err)
		return
	}

	if len(libs) == 0 {
		slog.Info("[FileWatcher] No libraries to watch")
		return
	}

	var watchRoots []string
	for _, lib := range libs {
		for _, p := range lib.Paths {
			if p == "" {
				continue
			}
			if _, err := os.Stat(p); err != nil {
				continue
			}
			if isRemoteMount(p) {
				slog.Info("[FileWatcher] Skipping remote mount", "path", p)
				continue
			}
			watchRoots = append(watchRoots, p)
		}
	}

	if len(watchRoots) == 0 {
		slog.Info("[FileWatcher] No local paths to watch")
		return
	}

	fw.mu.Lock()
	fw.stopCh = make(chan struct{})
	fw.running = true
	fw.mu.Unlock()
	stopCh := fw.stopCh

	go func() {
		watcher, err := fsnotify.NewWatcher()
		if err != nil {
			slog.Error("[FileWatcher] Failed to create watcher", "error", err)
			return
		}
		defer watcher.Close()

		watched := 0
		for _, p := range watchRoots {
			if err := addRecursive(watcher, p); err != nil {
				slog.Warn("[FileWatcher] Cannot watch", "path", p, "error", err)
			} else {
				watched++
			}
		}
		slog.Info("[FileWatcher] Watching paths for changes", "count", watched)

		for {
			select {
			case <-stopCh:
				slog.Info("[FileWatcher] Stop signal received")
				return
			case event, ok := <-watcher.Events:
				if !ok {
					return
				}
				fw.handle(watcher, event)
			case err, ok := <-watcher.Errors:
				if !ok {
					return
				}
				slog.Warn("[FileWatcher] Error", "error", err)
			}
		}
	}()
}

// handle 把一次 fsnotify 原始事件翻译成 1~2 条 IngestEvent。
// fsnotify 的 Op 是位掩码(Create|Write|Remove|Rename|Chmod),一次 Event 可能触发多种
// 语义,分开 Submit。
func (fw *FileWatcher) handle(watcher *fsnotify.Watcher, event fsnotify.Event) {
	if fw.ingest == nil && fw.refreshes == nil {
		return
	}
	path := event.Name
	now := time.Now()

	isDir := false
	if event.Has(fsnotify.Create) || event.Has(fsnotify.Write) || event.Has(fsnotify.Rename) {
		if info, err := os.Stat(path); err == nil {
			isDir = info.IsDir()
		}
	}

	if event.Has(fsnotify.Create) {
		if isDir {
			_ = addRecursive(watcher, path)
		}
		if !isDir && fw.refreshes != nil && classifySidecarPath(path) != "" {
			fw.refreshes.OnSidecarChange(path)
			return
		}
		fw.ingest.Submit(IngestEvent{
			Kind: EventCreate, Path: path, IsDir: isDir,
			Source: "fsnotify", DetectedAt: now,
		})
	}
	if event.Has(fsnotify.Write) {
		if !isDir && fw.refreshes != nil && classifySidecarPath(path) != "" {
			fw.refreshes.OnSidecarChange(path)
			return
		}
		fw.ingest.Submit(IngestEvent{
			Kind: EventModify, Path: path, IsDir: isDir,
			Source: "fsnotify", DetectedAt: now,
		})
	}
	if event.Has(fsnotify.Remove) {
		if fw.refreshes != nil && classifySidecarPath(path) != "" {
			fw.refreshes.OnSidecarChange(path)
			return
		}
		// 删除时文件已不存在,无法确定 isDir;processDelete 用 `= $1 OR LIKE $2/%`
		// 兜底目录级删除,与 isDir 无关。
		fw.ingest.Submit(IngestEvent{
			Kind: EventDelete, Path: path, IsDir: false,
			Source: "fsnotify", DetectedAt: now,
		})
	}
	if event.Has(fsnotify.Rename) {
		if fw.refreshes != nil && classifySidecarPath(path) != "" {
			fw.refreshes.OnSidecarChange(path)
			return
		}
		// fsnotify 的 Rename 语义:原路径被改名。新路径通常以 Create 紧随其后。
		// 跨目录 mv → 先 Rename(旧) 后 Create(新),renameBuffer 在 500ms 内合并成 Rename。
		// 若 stat 原路径还在(Windows 某些场景),按 Create 处理。
		if _, err := os.Stat(path); err == nil {
			fw.ingest.Submit(IngestEvent{
				Kind: EventCreate, Path: path, IsDir: isDir,
				Source: "fsnotify", DetectedAt: now,
			})
		} else {
			fw.ingest.Submit(IngestEvent{
				Kind: EventDelete, Path: path, IsDir: false,
				Source: "fsnotify", DetectedAt: now,
			})
		}
	}
}

func addRecursive(watcher *fsnotify.Watcher, path string) error {
	return filepath.Walk(path, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			name := filepath.Base(p)
			if strings.HasPrefix(name, ".") || strings.HasPrefix(name, "@") {
				return filepath.SkipDir
			}
			return watcher.Add(p)
		}
		return nil
	})
}

func (fw *FileWatcher) Stop() {
	fw.mu.Lock()
	defer fw.mu.Unlock()
	if fw.running && fw.stopCh != nil {
		close(fw.stopCh)
		fw.running = false
		slog.Info("[FileWatcher] Stopped")
	}
}

func (fw *FileWatcher) Restart(ctx context.Context, pool *pgxpool.Pool, cache *CacheService) {
	fw.Stop()
	fw.Start(ctx, pool, cache)
}

// isRemoteMount 检测路径是否落在远程挂载上,避免 fsnotify 去 watch 网盘(不支持且慢)。
// Unix: 调 df -T 解析 fstype;Windows: 检查 UNC 路径 + 通过 'net use' 判断映射盘。
func isRemoteMount(dirPath string) bool {
	if runtime.GOOS == "windows" {
		return isRemoteMountWindows(dirPath)
	}
	return isRemoteMountUnix(dirPath)
}

func isRemoteMountUnix(dirPath string) bool {
	cmd := exec.Command("df", "-T", dirPath)
	output, err := cmd.Output()
	if err != nil {
		return false
	}
	lines := strings.Split(string(output), "\n")
	if len(lines) < 2 {
		return false
	}
	parts := strings.Fields(lines[1])
	if len(parts) < 2 {
		return false
	}
	fsType := strings.ToLower(parts[1])
	remoteTypes := []string{"fuse", "nfs", "nfs4", "cifs", "smb", "smbfs", "9p", "sshfs"}
	for _, rt := range remoteTypes {
		if strings.HasPrefix(fsType, rt) || fsType == rt {
			return true
		}
	}
	return false
}

// isRemoteMountWindows 判定规则:
//  1. UNC 路径(\\server\share\... 或 //server/share/...)视为远程
//  2. 盘符形如 "Z:\..." → 用 'net use Z:' 判断是否映射网络驱动器
func isRemoteMountWindows(dirPath string) bool {
	if strings.HasPrefix(dirPath, `\\`) || strings.HasPrefix(dirPath, `//`) {
		return true
	}
	if len(dirPath) >= 2 && dirPath[1] == ':' {
		drive := strings.ToUpper(dirPath[:2])
		cmd := exec.Command("net", "use", drive)
		if err := cmd.Run(); err == nil {
			return true
		}
	}
	return false
}
